### ISSUE TITLE
LanguageSettingsPersister can be initialized on a background crashing package load

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#38130]
**Version Used**: 
Version 16.3.0 Preview 3.0 [29212.188.master]

**Steps to Reproduce**:

These are the steps I took:

1. Create a new C# Shared  Project
2. Open a CPS-based C# solution in the same session closing the above

**Expected Behavior**:
No crash

**Actual Behavior**:

```
Application: devenv.exe
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: Microsoft.VisualStudio.Composition.CompositionFailedException: An exception was thrown while initializing part "Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister". ---> System.InvalidOperationException: Unexpected false
   at Roslyn.Utilities.Contract.ThrowIfFalse(Boolean condition, String message)
   at Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject.AssertIsForeground()
   at Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject..ctor(IThreadingContext threadingContext, Boolean assertIsForeground)
   at Microsoft.VisualStudio.LanguageServices.Implementation.Options.LanguageSettingsPersister..ctor(IThreadingContext threadingContext, IServiceProvider serviceProvider, IGlobalOptionService optionService)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
   --- End of inner exception stack trace ---
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create()
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(PartLifecycleState nextState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveToState(PartLifecycleState requiredState)
   at Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose()
   at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.<>c__DisplayClass15_0.<GetExportedValueHelper>b__0()
   at Microsoft.VisualStudio.Composition.DelegateServices.<>c__DisplayClass2_0`1.<As>b__0()
   at System.Lazy`1.CreateValue()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Lazy`1.get_Value()
   at Microsoft.CodeAnalysis.Options.GlobalOptionService.LoadOptionFromSerializerOrGetDefault(OptionKey optionKey)
   at Microsoft.CodeAnalysis.Options.GlobalOptionService.GetOption(OptionKey optionKey)
   at Microsoft.CodeAnalysis.Options.OptionServiceFactory.OptionService.GetOption(OptionKey optionKey)
   at Microsoft.CodeAnalysis.Options.WorkspaceOptionSet.GetOption(OptionKey optionKey)
   at Microsoft.CodeAnalysis.Options.OptionSet.GetOption[T](Option`1 option)
   at Microsoft.VisualStudio.LanguageServices.Remote.RemoteHostClientServiceFactory.RemoteHostClientService.Enable()
   at Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2.EnableRemoteHostClientService()
   at Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2.<InitializeAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage.<InitializeAsync>d__2.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FailFast.OnFatalException(System.Exception)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.ErrorReporting.FatalError.ReportUnlessCanceled(System.Exception)
   at Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage+<InitializeAsync>d__2.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
   at Microsoft.VisualStudio.LanguageServices.CSharp.LanguageService.CSharpPackage+<InitializeAsync>d__2.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.AwaitTaskContinuation.InvokeAction(System.Object)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(System.Threading.ContextCallback, System.Object, System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.Run(System.Threading.Tasks.Task, Boolean)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task.FinishStageThree()
   at System.Threading.Tasks.Task.FinishStageTwo()
   at System.Threading.Tasks.Task.Finish(Boolean)
   at System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetException(System.Object)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Threading.Tasks.VoidTaskResult, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetException(System.Exception)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)
   at Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2+<InitializeAsync>d__8[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.AwaitTaskContinuation.InvokeAction(System.Object)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(System.Threading.ContextCallback, System.Object, System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.Run(System.Threading.Tasks.Task, Boolean)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task.FinishStageThree()
   at System.Threading.Tasks.Task`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetResult(Boolean)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetResult(Boolean)
   at Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2+<IsInIdeModeAsync>d__23[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.AwaitTaskContinuation.InvokeAction(System.Object)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(System.Threading.ContextCallback, System.Object, System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.Run(System.Threading.Tasks.Task, Boolean)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task.FinishStageThree()
   at System.Threading.Tasks.Task`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetResult(Boolean)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetResult(Boolean)
   at Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService.AbstractPackage`2+<IsInCommandLineModeAsync>d__24[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation+<>c.<.cctor>b__8_0(System.Object)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory+SingleExecuteProtector.TryExecute()
   at Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()
   at Microsoft.VisualStudio.Threading.JoinableTask.Join(System.Threading.CancellationToken)
   at Microsoft.VisualStudio.Threading.JoinableTask`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Join(System.Threading.CancellationToken)
   at Microsoft.VisualStudio.Shell.VsTaskLibraryHelper+<>c__DisplayClass33_1`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<AsVsTask>b__3()
   at Microsoft.VisualStudio.Shell.VsTaskLibraryHelper+<>c__DisplayClass33_0`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<AsVsTask>b__1(System.Object, Microsoft.VisualStudio.Shell.BlockingTaskEventArgs)
   at Microsoft.VisualStudio.Services.VsTask.RaiseOnMarkedAsBlockingEvent(Microsoft.VisualStudio.Shell.Interop.IVsTask)
   at Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](System.Func`1<System.__Canon>)
   at Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog(System.Action)
   at Microsoft.VisualStudio.Services.VsTask.InternalGetResult(Boolean)
   at Microsoft.VisualStudio.Services.VsTask.GetResult()
   at Microsoft.VisualStudio.Shell.Interop.IVsUIShellOpenDocument.OpenSpecificEditor(UInt32, System.String, System.Guid ByRef, System.String, System.Guid ByRef, System.String, Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchy, UInt32, IntPtr, Microsoft.VisualStudio.OLE.Interop.IServiceProvider, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.Open(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode, Boolean, UInt32, System.Guid ByRef, System.String, System.Guid ByRef, IntPtr, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.FileDocumentManager.OpenWithSpecific(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode, UInt32, System.Guid ByRef, System.String, System.Guid ByRef, IntPtr, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef, Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.WindowFrameShowAction)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode+<>c__DisplayClass490_0.<ReopenItem>b__0()
   at Microsoft.VisualStudio.ProjectSystem.VS.HResult.Invoke(System.Func`1<Microsoft.VisualStudio.ProjectSystem.VS.HResult>, System.IServiceProvider, Microsoft.VisualStudio.ProjectSystem.IProjectFaultHandlerService, Microsoft.VisualStudio.ProjectSystem.UnconfiguredProject)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.ReopenItem(UInt32, System.Guid ByRef, System.String, System.Guid ByRef, IntPtr, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef)
   at Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNodeWrapper.ReopenItem(UInt32, System.Guid ByRef, System.String, System.Guid ByRef, IntPtr, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef)
   at Microsoft.VisualStudio.Shell.Interop.IVsProject2.ReopenItem(UInt32, System.Guid ByRef, System.String, System.Guid ByRef, IntPtr, Microsoft.VisualStudio.Shell.Interop.IVsWindowFrame ByRef)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions.ReopenDocumentFrame(Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService, Microsoft.VisualStudio.Platform.WindowManagement.StubWindowInfo, Microsoft.VisualStudio.Platform.WindowManagement.ReopenDocumentOptions, Microsoft.VisualStudio.PlatformUI.Shell.DocumentGroup, Int32 ByRef)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions.ReopenDocumentFrame(Microsoft.VisualStudio.Platform.WindowManagement.IWindowManagerPrivate, Microsoft.VisualStudio.Platform.WindowManagement.StubWindowInfo, Microsoft.VisualStudio.Platform.WindowManagement.ReopenDocumentOptions, Microsoft.VisualStudio.PlatformUI.Shell.DocumentGroup, Int32 ByRef)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions.ReopenDocumentFrame(Microsoft.VisualStudio.Platform.WindowManagement.IWindowManagerPrivate, Microsoft.VisualStudio.Platform.WindowManagement.StubWindowInfo, Microsoft.VisualStudio.Platform.WindowManagement.ReopenDocumentOptions, Microsoft.VisualStudio.PlatformUI.Shell.DocumentGroup)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions+<>c__DisplayClass22_0.<LoadBinaryDocumentStates>b__0(Microsoft.VisualStudio.Utilities.VersionedBinaryReader)
   at Microsoft.VisualStudio.Utilities.VersionedBinaryReader.ReadVersioned(Int32, ReadCallback, Boolean)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions.LoadBinaryDocumentStates(Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService, Microsoft.VisualStudio.OLE.Interop.IStream)
   at Microsoft.VisualStudio.Platform.WindowManagement.SolutionOptions.LoadDocumentStates(Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService, Microsoft.VisualStudio.OLE.Interop.IStream)
   at Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService+<>c__DisplayClass193_0.<LoadDocumentStates>b__0()
   at Microsoft.VisualStudio.ErrorHandler.CallWithCOMConvention(System.Action, Boolean)
   at Microsoft.VisualStudio.Platform.WindowManagement.WindowManagerService.LoadDocumentStates(Microsoft.VisualStudio.OLE.Interop.IStream, Int32)
   at Microsoft.VisualStudio.Shell.Interop.IVsSolution.OpenSolutionFile(UInt32, System.String)
   at Microsoft.VisualStudio.PlatformUI.Packages.MRU.ExtensionMethods.TryOpenProject(Microsoft.VisualStudio.Shell.Interop.IVsSolution, System.String, System.Guid)
   at Microsoft.VisualStudio.PlatformUI.CodeContainerAccessManager+<OpenLocalCodeContainerAsync>d__9.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.AwaitTaskContinuation.InvokeAction(System.Object)
   at System.Threading.Tasks.AwaitTaskContinuation.RunCallback(System.Threading.ContextCallback, System.Object, System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation.Run(System.Threading.Tasks.Task, Boolean)
   at System.Threading.Tasks.Task.FinishContinuations()
   at System.Threading.Tasks.Task.FinishStageThree()
   at System.Threading.Tasks.Task`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].TrySetResult(System.__Canon)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].SetResult(System.__Canon)
   at Microsoft.VisualStudio.PlatformUI.CodeContainerAccessManager+<GetSolutionAsync>d__3.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.InvokeMoveNext(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore+MoveNextRunner.Run()
   at System.Threading.Tasks.SynchronizationContextAwaitTaskContinuation+<>c.<.cctor>b__8_0(System.Object)
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory+SingleExecuteProtector.TryExecute()
   at Microsoft.VisualStudio.Threading.JoinableTaskFactory+SingleExecuteProtector+<>c.<.cctor>b__20_0(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(System.Object)
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)

```

This component is a MEF component but it has UI thread affinity: http://index/?query=LanguageSettingsPersister&rightProject=Microsoft.VisualStudio.LanguageServices&file=Implementation%5COptions%5CLanguageSettingsPersister.cs&line=28. These concepts don't mix.

**Comment by jasonmalinowski** [comment]
Not able to reproduce, and the stack given really doesn't make sense. Going to close.

**Comment by jasonmalinowski** [comment]
If anybody hits this and has a dump, I'd love to look!

**Comment by davkean** [comment]
@jasonmalinowski Maybe you are looking the wrong way - maybe what it thinks is the UI thread, isn't the UI thread?

**Comment by jasonmalinowski** [comment]
@davkean I did think about that, and concluded that if it was happening, I need a dump to investigate. :smile: At this point we just use the JTF's concept of what the UI thread is (all the old junk is gone), so something is really wrong...

**Comment by davkean** [comment]
I have a dump, just uploading it.

**Comment by davkean** [comment]
Dump will in [internalshare]\public\davkean\Roslyn\38130.zip.

**Comment by jasonmalinowski** [comment]
My belief is this is fixed by #38356 -- in there something was causing an initialization to happen on a background thread first, causing the MEF part to fail, and I'm _guessing_ MEF then cached the exception and rethew it when the UI thread also touched that component. HeeJae's change added a SwitchToMainThread() which implicitly fixes this.

Leaving open while I'm confirming with the VS MEF team the assumption I'm making about it's behavior.

**Comment by davkean** [comment]
Instead of relying on ordering - why don't you change the MEF component so that it can be initialized free-threaded and then do its UI-thread-bound initialization on first usage?

**Comment by jasonmalinowski** [comment]
We're just going to delete the code entirely and remove the use of the VS service; as there's a lot more problems with this all than just the UI thread affinity.

**Comment by jasonmalinowski** [comment]
Anyways the MEF team confirmed that indeed exceptions are rethrown, so this is fixed by #38356. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/639090 is tracking the longer-term cleanup of this entire system.
